arch: arm: Add hook called when exiting CPU idle state

nordic-krch opened the pull request:
Added a hook which is called when CPU exits cpu idle. There is already one optional which is called on entering CPU idle. Those two can be used to track CPU activity.

There is already a macro hook which is called immediately after the wake up and it is used to inject specific code where primary use has was HW errata. This hook is more generic since it is a C function which user can implement.

ithinuel reviewed (commented):
Is my understanding correct that this change only saves users from having to create the `soc_cpu_idle.h` file containing (at least) the following:
```c
#define SOC_ON_EXIT_CPU_IDLE z_arm_on_exit_cpu_idle();
```

nordic-krch commented:
@ithinuel 
> Is my understanding correct that this change only saves users from having to create the soc_cpu_idle.h file containing (at least) the following

Macro approach was added to handle hw workaround in one SoC where certain code had to be executed as the first one after wake up. Callback approach is more flexible. It would be hard to use more generic feature (like CPU load measurement) in SoC that already implements macro for a workaround.

I would like to be able to measure CPU load in a user-friendly way. With macro approach if user would like to enable CPU load measuring he would need to add custom header, include path in cmake, not user friendly. Do you have any suggestions?

ithinuel commented:
Oh, so the macro-hook is for soc-vendor’s workarounds and the function-hook is for user-code?
That’d make sense to have them both then (different audience and implementation points).

Could their name & documentation express that intent better then please? (just to reduce the chance of user using the macros and soc-vendor adding the workaround using the C function).

nordic-krch committed:
arch: arm: Add hook called when exiting CPU idle state

Added a hook which is called when CPU exits cpu idle. There is already
one optional which is called on entering CPU idle. Those two can be
used to track CPU activity.

There is already a macro hook which is called immediately after the
wake up and it is used to inject specific code where primary use has
was HW errata. This hook is more generic since it is a C function which
user can implement.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@ithinuel regarding Kconfig options I already added a note to `ARM_ON_EXIT_CPU_IDLE` to use the function hook for more generic operations and I named the option for function hook  `ARM_ON_EXIT_CPU_IDLE_HOOK` (since on enter hook is named `ARM_ON_ENTER_CPU_IDLE_HOOK`. Can you suggest what naming you had in mind?

nordic-krch commented:
@ithinuel ping regarding naming change suggestions.

ithinuel commented:
@nashif is there a convention for "arch-facing" API/Hook vs "user-facing" API/Hook ?

nordic-krch commented:
@nashif can you help here?

bjarki-andreasen commented:
I would love for there to be a proper CPUI exit handle to be used for PM as well, currently we have quite "hardcoded" solutions like:
https://github.com/zephyrproject-rtos/zephyr/blob/c36b3c02969894e5d953b0978253150471c3c801/arch/arm/core/cortex_m/irq_manage.c#L132-L135
which would definitely be improved substantially with a generic `soc_idle_exit()` for PM and tracing to use.

It does not solve a potential free to use "user hook" for idle exit (though that could be put into `soc_idle_exit()` as well), but I don't think that's actually something we want to expose that high level anyway :)

wearyzen commented:
Hi @nordic-krch,
Tracking CPU utilization is a common requirement, so the new API should be platform-agnostic to ensure ease of use. The sys_trace_* framework in Zephyr appears to be a generic and suitable choice for this purpose. 

Currently, sys_trace_idle() is called at the start of both arch_cpu_idle and arch_cpu_atomic_idle. Introducing a new API, such as sys_trace_idle_exit(), at the end of these functions could achieve the same result without tying the implementation to ARM-specific APIs (which to me seem more related to power management than tracing).

wearyzen commented:
> Hi @nordic-krch, Tracking CPU utilization is a common requirement, so the new API should be platform-agnostic to ensure ease of use. The sys_trace_* framework in Zephyr appears to be a generic and suitable choice for this purpose.
> 
> Currently, sys_trace_idle() is called at the start of both arch_cpu_idle and arch_cpu_atomic_idle. Introducing a new API, such as sys_trace_idle_exit(), at the end of these functions could achieve the same result without tying the implementation to ARM-specific APIs (which to me seem more related to power management than tracing).

Hi @nordic-krch, any thoughts on this?

nordic-krch commented:
Arch agnostic option is tempting. However, I would like to be able to get API like `get_cpu_load` and I might not want to turn on tracing for that since that will increase the load with tracing hooks all over the place. It might also be possible that there is no available tracing backend on the target. Maybe we could have minimal tracing backend that just implements `sys_trace_idle` and `sys_trace_idle_exit` that could be used for cpu load api.

wearyzen commented:
> Arch agnostic option is tempting. However, I would like to be able to get API like `get_cpu_load` and I might not want to turn on tracing for that since that will increase the load with tracing hooks all over the place. It might also be possible that there is no available tracing backend on the target. Maybe we could have minimal tracing backend that just implements `sys_trace_idle` and `sys_trace_idle_exit` that could be used for cpu load api.

I agree with the concerns and like you suggested we can have a minimal tracing backend just for capturing CPU load. This would still be useful since it can be reused by other archs. We might even suggest a config flag to just enable these 2 APIs in existing tracing backend.
Let me drop a message in the tracing channel and see what others think about this and if someone has a better option.

nordic-krch commented:
Closing in favor of #85131
